### PR TITLE
[Assist] Provide interactive updates during agent execution

### PR DESCRIPTION
**Pull request by xacrimon**
Partially resolves [#1524](https://github.com/gravitational/teleport.e/issues/1524)

### CODE REVIEW

**Comment by xacrimon** [comment]
fyi: looking for review on code & architecture of changes. pushing the updated streaming tests in a moment once I have them all fixed so they pass.

**Review by ryanclark — APPROVED** [review]

File: lib/web/assistant.go [hunk 1]
```diff
@@ -422,7 +423,11 @@ func runAssistant(h *Handler, w http.ResponseWriter, r *http.Request,
 	}
 
 	// onMessageFn is called when a message is received from the OpenAI API.
+	var writeMu sync.Mutex
 	onMessageFn := func(kind assist.MessageType, payload []byte, createdTime time.Time) error {
+		writeMu.Lock()
+		defer writeMu.Unlock()
```
Comment (jakule) [inline lib/web/assistant.go:429]: Why do we need a mutex here? The socket is already serialized.

**Review by jakule — COMMENTED** [review]
1. This PR needs some tests
2. I'd consider a different approach when dealing with channels or replace them with a callback function.

File: lib/ai/model/agent.go [hunk 2]
```diff
@@ -307,34 +335,49 @@ func parseJSONFromModel[T any](text string) (T, *invalidOutputError) {
...
 
 // parsePlanningOutput parses the output of the model after asking it to plan it's next action
 // and returns the appropriate event type or an error.
-func parsePlanningOutput(text string) (*agentAction, *agentFinish, error) {
+func parsePlanningOutput(deltas <-chan string) (*AgentAction, *agentFinish, error) {
+	var text string
+	for delta := range deltas {
+		text += delta
+
+		if strings.HasPrefix(text, "<FINAL RESPONSE>") {
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply lib/ai/model/agent.go:348]: yup

File: lib/ai/model/agent.go [hunk 3]
```diff
@@ -163,39 +171,43 @@ func (a *Agent) takeNextStep(ctx context.Context, state *executionState) (stepOu
 
 	// If finish is set, the agent is done and did not call upon any tool.
 	if finish != nil {
+		close(progressUpdates)
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply lib/ai/model/agent.go:175]: Replaced with callback

File: lib/ai/chat.go [hunk 4]
```diff
@@ -56,7 +56,7 @@ func (chat *Chat) GetMessages() []openai.ChatCompletionMessage {
 // Message types:
 // - CompletionCommand: a command from the assistant
 // - Message: a text message from the assistant
-func (chat *Chat) Complete(ctx context.Context, userInput string) (any, error) {
+func (chat *Chat) Complete(ctx context.Context, userInput string, progressUpdates chan<- *model.AgentAction) (any, error) {
```
[the remark this answers is not in the document]
Comment (xacrimon) [reply lib/ai/chat.go:59]: Replaced with callback

File: lib/ai/model/agent.go [hunk 5]
```diff
@@ -28,10 +31,11 @@ import (
...
 const (
-	actionFinalAnswer = "Final Answer"
-	actionException   = "_Exception"
-	maxIterations     = 15
-	maxElapsedTime    = 5 * time.Minute
+	actionFinalAnswer   = "Final Answer"
+	actionException     = "_Exception"
+	maxIterations       = 15
+	maxElapsedTime      = 5 * time.Minute
+	finalResponseHeader = "<FINAL RESPONSE>"
```
Comment (jakule) [inline lib/ai/model/agent.go:38]: Can you please add Go docs here? These constants are private, but I don't think they are self-explanatory, so adding comments can make the code easier to understand.

File: lib/ai/model/agent.go [hunk 6]
```diff
@@ -307,35 +329,50 @@ func parseJSONFromModel[T any](text string) (T, *invalidOutputError) {
...
+	Reasoning    string `json:"reasoning"`
 }
 
 // parsePlanningOutput parses the output of the model after asking it to plan it's next action
 // and returns the appropriate event type or an error.
-func parsePlanningOutput(text string) (*agentAction, *agentFinish, error) {
+func parsePlanningOutput(deltas <-chan string) (*AgentAction, *agentFinish, error) {
+	var text string
+	for delta := range deltas {
+		text += delta
```
Comment (jakule) [inline lib/ai/model/agent.go:340]: nit: `StringBuilder`

**Review by jakule — APPROVED** [review]
LGTM. The PR needs to be rebased to fix the tests/compilation.

File: lib/ai/model/agent.go (hunk 5 above)
Comment (xacrimon) [reply lib/ai/model/agent.go:38]: Solid

File: lib/ai/model/agent.go (hunk 6 above)
Comment (xacrimon) [reply lib/ai/model/agent.go:340]: This is actually slower here than a plain `string` due to us needing to perform computation on the entire string every iteration. `StringBuilder` only pays off if you do more writes than anything else and this isn't the case here. It incurs nontrivial overhead similar to an append when you call `builder.String()`

**Pull request merged**